### PR TITLE
Add scribing question controller

### PR DESCRIPTION
**Pull request by nicholette**
PR to be merged before scribing question React frontend.

### CODE REVIEW

File: app/controllers/course/assessment/question/scribing_controller.rb [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    render json: { message: message },
+           status: :ok
+  end
+
+  def render_failure_json(message)
+    render json: { message: message, errors: @scribing_question.errors.full_messages },
+           status: :bad_request
+  end
+
+  def generate_pdf_questions(file) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
```
Comment (weiqingtoh) [inline app/controllers/course/assessment/question/scribing_controller.rb]: I think you can move this into a service. Easier to manage and refactor when the logic is in a separate file.

File: app/controllers/course/assessment/question/scribing_controller.rb [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+  def show
+    respond_to do |format|
+      format.json { render_scribing_question_json }
+    end
+  end
+
+  def create # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+    # special handling for PDF files
+    if params&.[](:question_scribing)&.[](:file)&.content_type&.downcase == 'application/pdf'
```
Comment (weiqingtoh) [inline app/controllers/course/assessment/question/scribing_controller.rb]: I just realised Ruby has a `dig` function - http://ruby-doc.org/core-2.3.0_preview1/Hash.html#method-i-dig

Alternatively, you can place this in a method, and call this `if file_for_uploading?`, and extend when we support other file types.

File: app/controllers/course/assessment/question/scribing_controller.rb [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+  def edit
+    respond_to do |format|
+      format.html { render 'edit' }
+      format.json { render_scribing_question_json }
+    end
+  end
+
+  def update
+    # Don't allow the user to edit attachment at the moment
```
Comment (weiqingtoh) [inline app/controllers/course/assessment/question/scribing_controller.rb]: Can put this as a TODO instead. If need be, you can also guard against this in params by detecting the controller action. , but I think this would suffice for now.

Comment (weiqingtoh) [inline app/controllers/course/assessment/question/scribing_controller.rb]: Anyway, how do you plan to handle this case in the future?

File: spec/factories/course_assessment_question_scribings.rb [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # frozen_string_literal: true
 FactoryGirl.define do
-  factory :course_assessment_question_scribing,
+  factory :course_assessment_question_scribings,
```
Comment (weiqingtoh) [inline spec/factories/course_assessment_question_scribings.rb]: Singular should be the correct convention, why did you change?

File: app/models/course/assessment.rb [hunk 5]
```diff
@@ -40,6 +40,9 @@ class Course::Assessment < ActiveRecord::Base
   has_many :programming_questions,
            through: :questions, inverse_through: :question, source: :actable,
            source_type: Course::Assessment::Question::Programming.name
+  has_many :scribing_questions,
```
Comment (weiqingtoh) [inline app/models/course/assessment.rb]: Can you shift this and the abilities into a separate commit, then controllers + views go into 1 commit?

File: app/controllers/course/assessment/question/scribing_controller.rb (hunk 3 above)
Comment (nicholette) [inline app/controllers/course/assessment/question/scribing_controller.rb]: This case needs more clarification on the behaviour before proceeding. 
The troublesome behaviour is when someone edits a file with a multipage pdf => does that mean creating more questions (same as the behaviour for create), or picking up a certain page from the pdf.

File: app/views/course/assessment/question/scribing/_scribing_question.json.jbuilder [hunk 6]
```diff
@@ -0,0 +1,21 @@
+json.question do
+  json.(@scribing_question, :id, :title, :description, :staff_only_comments, :maximum_grade,
+    :weight)
+  
+  if @scribing_question.attachment_references.present?
+    json.attachment_reference do
```
Comment (weiqingtoh) [inline app/views/course/assessment/question/scribing/_scribing_question.json.jbuilder]: Factor this out into an attachment reference jbuilder partial, and reuse that.

**Comment by weiqingtoh** [comment]
Pls also fix newlines!

File: spec/factories/course_assessment_question_scribings.rb (hunk 4 above)
Comment (nicholette) [inline spec/factories/course_assessment_question_scribings.rb]: oops. I was following the file name ><

File: app/views/course/assessment/question/scribing/_scribing_question.json.jbuilder (hunk 6 above)
Comment (nicholette) [inline app/views/course/assessment/question/scribing/_scribing_question.json.jbuilder]: ok

File: app/controllers/course/assessment/question/scribing_controller.rb (hunk 3 above)
Comment (weiqingtoh) [inline app/controllers/course/assessment/question/scribing_controller.rb]: I think once there are scribbles and submissions we shouldn't be allowing changes to the question, so any updates should only happen to the meta-data. If the staff wants to change the underlying PDF/image, he/she should delete this question and create a new one. 

Thoughts? @allenwq

File: README.md [hunk 7]
```diff
@@ -16,11 +16,12 @@ educators to increase student engagement and make learning fun.
...
 1. Ruby (>= 2.3.1)
-2. Ruby on Rails
-3. PostgreSQL
-4. ImageMagick or GraphicsMagick (For [MiniMagick](https://github.com/minimagick/minimagick))
-5. Node.js
-6. Yarn
+1. Ruby on Rails
+1. PostgreSQL
+1. ImageMagick or GraphicsMagick (For [MiniMagick](https://github.com/minimagick/minimagick))
+1. Ghostscript (for PDF to image processing for [MiniMagick](https://github.com/minimagick/minimagick))
```
Comment (allenwq) [inline README.md]: Is this a dependency of ImageMagick ?  http://imagemagick.org/script/index.php

Comment (weiqingtoh) [inline README.md]: Imagemagick deals only with images. However they have 'drivers' or 'sub-libraries' that helps do cross file-type conversion. Ghostscript helps to provide imagemagick functionality for pdf, so we need it for the scribing question. 

This took me a while to debug on my local machine cause imagemagick didn't have error messages for this >_< 

Surprisingly travis has this already, but we will need to ensure that our staging and production dependencies are updated for this as well. 

cc: @fonglh

File: app/services/course/assessment/question/scribing_import_service.rb [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+  # Imports and saves the provided PDF as a scribing question.
+  #
+  # @return [Boolean] True if the pdf is processed and successfully saved, otherwise false. Note
+  #   that if the save is unsuccessful, all questions are not persisted.
+  def save
+    return_value = true
+    Course::Assessment::Question::Scribing.transaction do
+      generate_pdf_files.each do |file|
+        question = build_scribing_question
```
Comment (allenwq) [inline app/services/course/assessment/question/scribing_import_service.rb]: Not very understand why the service takes a question as the param and creates a new question here instead, it's very confusing because the user would expect the question given to be saved. 

I'd think it's better to remove the question from param and let the service to take only params from the controller and return a created question.

Comment (weiqingtoh) [inline app/services/course/assessment/question/scribing_import_service.rb]: That works too, but if we pass in params, there is some coupling with the controllers params structure, so specifying `question` and `file` here would make this independent.

File: README.md (hunk 7 above)
Comment (allenwq) [inline README.md]: Actually, the scribing tests passed on my local also.  Maybe merge this with the line above ? cause it's imagemagick specific.

**Comment by allenwq** [comment]
LGTM otherwise
